src/main/java/gov/usgs/earthquake/nats/NATSClient.java

```diff
@@ -0,0 +1,165 @@
+ * Written so several concurrent connections can be created.
+ */
+public class NATSClient implements Configurable {
+
+  public static Logger LOGGER  = Logger
+          .getLogger(NATSClient.class.getName());
+
+  public static String SERVER_HOST_PROPERTY = "serverHost";
+  public static String SERVER_PORT_PROPERTY = "serverPort";
+  public static String CLUSTER_ID_PROPERTY = "clusterId";
+  public static String CLIENT_ID_PROPERTY = "clientId";
+  public static String SUBJECT_PROPERTY = "subject";
+
+  private String serverHost;
+  private String serverPort;
+  private String clusterId;
+  private String clientId;
+  private String subject;
+
+  private StreamingConnection connection;
+
+  /**
+   * Configures the required and optional parameters to connect to NATS Streaming server
+   *
+   * @param config
+   *            the Config to load.
+   * @throws Exception
+   */
+  @Override
+  public void configure(Config config) throws Exception {
+    // required parameters
+    serverHost = config.getProperty(SERVER_HOST_PROPERTY);
+    serverPort = config.getProperty(SERVER_PORT_PROPERTY);
+    clusterId = config.getProperty(CLUSTER_ID_PROPERTY);
+
+    clientId = config.getProperty(CLIENT_ID_PROPERTY);
+    if (clientId == null) {
+      clientId = generateClientId();
+    }
+    subject = config.getProperty (SUBJECT_PROPERTY); //make optional (provide default)
+  }
+
+  /**
+   * Starts connection to NATS streaming server
+   * @throws Exception If something goes wrong when connecting to NATS streaming server
+   */
+  @Override
+  public void startup() throws Exception {
+    // create connection
+    StreamingConnectionFactory factory = new StreamingConnectionFactory(clusterId,clientId);
+    factory.setNatsUrl("nats://" + serverHost + ":" + serverPort);
+    connection = factory.createConnection();
+  }
+
+  /**
+   * Safely closes connection to NATS Streaming server
+   */
+  @Override
+  public void shutdown() {
+    try {
+      connection.close();
+    } catch (Exception e) {
+      LOGGER.log(Level.WARNING, "[" + getName() + "] failed to close NATS connection");
+    }
+    connection = null;
+  }
+
+  /**
+   * Creates a client ID based on the host IP and MAC address
+   *
+   * @return clientId
+   * @throws Exception if there's an issue accessing IP or MAC addresses, or can't do sha1 hash
+   */
+  private static String generateClientId() throws Exception {
+    // get mac address
+    InetAddress host = InetAddress.getLocalHost();
+    NetworkInterface net = NetworkInterface.getByInetAddress(host);
+    byte[] macRaw =  net.getHardwareAddress();
+
+    // do a sha1 hash
+    MessageDigest digest = MessageDigest.getInstance("SHA-1");
+    digest.reset();
+    digest.update(macRaw);
+    String sha1 = String.format("%040x", new BigInteger(1, digest.digest()));
+
+    // create client id
+    String clientId = host.getHostAddress().replace('.','-') + '_' + sha1+ '_' + Thread.currentThread().getId();
```

Comment:
I would make the class take a parameter.  You can default it to thread id (converted to string), but there may otherwise be collisions across processes...  configured pdl components can use getName() as a unique-ish value that is consistent across process starts

Reply:
On second thought, this static method is useful on it's own without the Thread id.  Other classes can use this to get a unique-ish id if needed and append their own sequence.

Reply:
Updated the NATSClient class to optionally take a suffix parameter on construction, and generateClientId to take a suffix as well. If NATSClient isn't constructed with a suffix string, it uses the thread ID.